### PR TITLE
Swapped usage of rosidl_cmake over to the new rosidl_pycommon.

### PR DESCRIPTION
**Pull request by maspe36**
As of [rosidl 3.3.0](https://github.com/ros2/rosidl/commit/9348ce9b466335590dc334aab01f4f0dd270713b) (specifically, this [commit](https://github.com/ros2/rosidl/pull/696/files#diff-330d15435e7274b3796bb084ab591a6be34c8afeb8acdcc90e93074cc33bcc12)), the rosidl_cmake Python module was moved to a new rosidl_pycommon package and the Python module in rosidl_cmake was deprecated.

This was throwing these deprecation warnings when building for rolling. 
```
Starting >>> builtin_interfaces
Starting >>> unique_identifier_msgs
--- stderr: unique_identifier_msgs                                                                                 
/opt/ros/rolling/local/lib/python3.10/dist-packages/rosidl_cmake/__init__.py:19: UserWarning: The 'rosidl_cmake' Python module is deprecated. Use 'rosidl_pycommon' instead.
  warnings.warn(
---
Finished <<< unique_identifier_msgs [2.73s]
--- stderr: builtin_interfaces                               
/opt/ros/rolling/local/lib/python3.10/dist-packages/rosidl_cmake/__init__.py:19: UserWarning: The 'rosidl_cmake' Python module is deprecated. Use 'rosidl_pycommon' instead.
  warnings.warn(
---
Finished <<< builtin_interfaces [2.92s]
```

### CODE REVIEW

File: rosidl_generator_rs/package.xml
Comment (maspe36) [inline rosidl_generator_rs/package.xml:16]: I may be interpreting [REP 149](https://www.ros.org/reps/rep-0149.html#id20) incorrectly, but to me it reads as if I should be able to use a less/greater than for ROS distros. Unfortunately that doesn't work, and the xml fails to parse. Not sure what to do if rosidl_cmake 3.3.0+ gets back ported to Humble or when Iron comes out

Comment (nnmm) [reply rosidl_generator_rs/package.xml:16]: Reads that way to me too. Did you try only `>` or also `>=`?

**Comment by nnmm** [comment]
This would fit the bill for `package.xml`, right?

```
<buildtool_export_depend condition="$ROS_DISTRO == foxy or $ROS_DISTRO == galactic or $ROS_DISTRO == humble">rosidl_cmake</buildtool_export_depend>
<buildtool_export_depend condition="$ROS_DISTRO != foxy and $ROS_DISTRO != galactic and $ROS_DISTRO != humble">rosidl_pycommon</buildtool_export_depend>
```

Then in CMake we could do something a little more elegant like @esteve proposed.

File: rosidl_generator_rs/package.xml
Comment (maspe36) [reply rosidl_generator_rs/package.xml:16]: Coming back to this now, looks to only be a problem with less than operators.

Fails to parse 
`$ROS_DISTRO <= rolling`
`$ROS_DISTRO < rolling`

Successfully parses 
`$ROS_DISTRO >= rolling`
`$ROS_DISTRO > rolling`

Thankfully this means we can handle this a bit more concisely in the `package.xml`

```xml
<buildtool_export_depend condition="rolling > $ROS_DISTRO">rosidl_cmake</buildtool_export_depend>
<buildtool_export_depend condition="$ROS_DISTRO >= rolling">rosidl_pycommon</buildtool_export_depend>
```
And then we can swap the usage of `rolling` to `iron` once its out.

**Comment by maspe36** [comment]
> The changes assume that rolling is the next ROS distro, however it's more akin to main in a GitHub repo that gets tagged and release, so these changes won't work with the next ROS distro iron

The next patch set doesn't assume this and I don't think will break when `iron` is released.

> create a new branch for humble and make main target iron and greater, and backport any changes to humble, galactic, etc.

I think eventually this may need to happen, but for now that's probably more work than its worth. Especially since we can have optional dependencies in the `package.xml` based on distro.

> check the version of one of the rosidl base packages in CMake (e.g. rosidl_parser) and if it's greater or equal than 3.3.0, then don't call find_package on rosidl_cmake

Is this really better than doing the same check from the `package.xml` in the CMake?

```cmake
if("$ENV{ROS_DISTRO}" STRLESS_EQUAL "humble")
    find_package(rosidl_cmake REQUIRED)
    ament_export_dependencies(rosidl_cmake)
endif()
```
Only benefit I see would be if it gets backported but since this is a deprecation I think @nnmm is right, this probably won't get backported.

**Comment by maspe36** [comment]
Look like the build is failing for the same reason mentioned [here](https://github.com/ros2-rust/ros2_rust/pull/300)

**Review by nnmm — DISMISSED** [review]
From my understanding, this looks good. I'll wait a little bit before merging to give others the chance to take a look too.

File: rosidl_generator_rs/package.xml
Comment (esteve) [inline rosidl_generator_rs/package.xml:20]: Please add `buildtool_depend` and `buildtool_export_depend` for `ros_environment`, which provides access to the `ROS_DISTRO` environment variable. Although usually when building locally `ROS_DISTRO` is available, it's not guaranteed unless `ros_environment` is added as a dependency.

**Review by esteve — CHANGES_REQUESTED** [review]

File: rosidl_generator_rs/package.xml
Comment (maspe36) [reply rosidl_generator_rs/package.xml:22]: Done

**Comment by nnmm** [comment]
@maspe36 Not quite "tonight" as promised, but I have an MR open with the fix, just waiting for review by a maintainer: https://github.com/ros2-rust/ros2_rust/pull/309

**Comment by nnmm** [comment]
@maspe36 If you rebase this onto main, CI should be green.

**Review by nnmm — COMMENTED** [review]
Don't we also need `rosidl_pycommon` in `rosidl_generator_rs/CMakeLists.txt` too? Otherwise LGTM, but I can't merge before @esteve revisits his requested changes.

**Comment by maspe36** [comment]
@esteve @nnmm [rosidl_pycommon](https://github.com/ros2/rosidl/tree/rolling/rosidl_pycommon) is a python only package, you cannot find it with `find_package`, and you cannot export it with `ament_export_dependencies`. Having it in the `package.xml` should be sufficient for rosdep

**Review by nnmm — APPROVED** [review]
Sounds good to me. Thanks for taking care of this, and sorry it took forever to merge!

**Review by esteve — APPROVED** [review]

**Comment by esteve** [comment]
@maspe36 thanks again for all the patience! :slightly_smiling_face:

**Pull request merged**
